### PR TITLE
Port away from Distribution.Text and Distribution.Compat.ReadP

### PR DESCRIPTION
**Pull request by solpeth**
`Distribution.Text` and `Distribution.Compat.ReadP` are deprecated, and in Cabal-3 the former has its typeclass entirely removed and the latter is removed entirely (or, more accurately, moved to a `Deprecated` module).

To aid in the eventual migration to Cabal-3 inside hackport, I've hacked away at porting our use of `Distribution.Text` to mainly two, separate modules: `Distribution.Pretty` for pretty printing, and `Distribution.Parsec.Class` for parsing. I've also used `Distribution.Compat.CharParsing` for helpful character parsing functions that are seemingly a drop-in replacement for the `Distribution.Compat.ReadP` parsing functions.

I have noticed no regressions with simple `hackport merge x` commands so far.

While this PR is fine to be merged as-is provided the design and implementation is okay with @trofi and others, `Portage.PackageId` still relies on `readP_to_S` to run the parser in the `parseFriendlyPackage` function. This is fine for the moment, but it will be a blocker to the eventual migration to Cabal-3 under the hood.

I'm happy to keep hacking away at a solution (which as far as I currently understand will use something from `Text.Parsec`) but I'm all ears to some guidance on this one, as I'm not all that familiar with Parsec.

### CODE REVIEW

**Comment by solpeth** [comment]
@trofi Thank you! I'll see if I can port `parseFriendlyPackage` in the near future and merge it all together.

Here is the latest attempt at `parseFriendlyPackage`:

```haskell
parseFriendlyPackage :: String -> Maybe (Maybe Category, Cabal.PackageName, Maybe Portage.Version)
parseFriendlyPackage str =
  case runParsecParser parser "<parseFriendlyPackage>" (fieldLineStreamFromString str) of
    Right x -> Just x
    _ -> Nothing
  where
  parser = do
    mc <- P.optional $ do
      c <- parsec
      _ <- P.char '/'
      return c
    p <- parsec
    mv <- P.optional $ do
      v <- parsec
      return v
    return (mc, p, mv)
```

`runParsecParser` is from `Distribution.Parsec.Class`. This will build, but parsing will always fail. I'll hack away and see what happens.

**Comment by trofi** [comment]
> @trofi Thank you! I'll see if I can port `parseFriendlyPackage` in the near future and merge it all together.
> 
> Here is the latest attempt at `parseFriendlyPackage`:
> 
> ```haskell
> parseFriendlyPackage :: String -> Maybe (Maybe Category, Cabal.PackageName, Maybe Portage.Version)
> parseFriendlyPackage str =
>   case runParsecParser parser "<parseFriendlyPackage>" (fieldLineStreamFromString str) of
>     Right x -> Just x
>     _ -> Nothing
>   where
>   parser = do
>     mc <- P.optional $ do
>       c <- parsec
>       _ <- P.char '/'
>       return c
>     p <- parsec
>     mv <- P.optional $ do
>       v <- parsec
>       return v
>     return (mc, p, mv)
> ```
> 
> `runParsecParser` is from `Distribution.Parsec.Class`. This will build, but parsing will always fail. I'll hack away and see what happens.

Ah, parsec does only 1-char backtracking (as opposed to `ReadP`'s full backtracking). You might need to insert parsec's `try` at a backtracking point, like `P.optional $ do` -> `P.optional $ P.try $ do` (unless `optional` is supposed to do the same). You can add a small test to `hackport.cabal` and i'll be able to look in more detail.

**Comment by solpeth** [comment]
Thank you, I'll give that a try.

**Comment by solpeth** [comment]
That worked. Thanks again for the direction! I'll push the change for review and comment.

File: Portage/PackageId.hs [hunk 1]
```diff
@@ -109,64 +123,36 @@ toCabalPackageId (PackageId (PackageName _cat name) version) =
...
-    version <- parse
-    return (PackageId name version)
-
 parseFriendlyPackage :: String -> Maybe (Maybe Category, Cabal.PackageName, Maybe Portage.Version)
 parseFriendlyPackage str =
-  case [ p | (p,s) <- Parse.readP_to_S parser str
-       , all Char.isSpace s ] of
-    [] -> Nothing
-    (x:_) -> Just x
+  case explicitEitherParsec parser str of
```
Comment (solpeth) [inline Portage/PackageId.hs]: This parser is failing to parse correctly if a version is specified in the package string to be parsed.

**Comment by solpeth** [comment]
Update: the new parser currently works unless a version is specified in the package string to be parsed. I'll keep looking into where it's going wrong.

```haskell
 case explicitEitherParsec parser str of
    Right x -> Just x
    Left _ -> Nothing
  where
  parser = do
    mc <- P.optional . P.try $ do
      c <- parsec
      _ <- P.char '/'
      return c
    p <- parsec
    mv <- P.optional $ do
      _ <- P.char '-'
      v <- parsec
      return v
    return (mc, p, mv)
```

**Comment by solpeth** [comment]
It looks like the parser is failing at `p <- parsec`, which should parse a Cabal `PackageName`. The underlying `Distribution.Parsec.Class.parsecUnqualComponentName` function call returns with the error message "unexpected all digits in portion of unqualified component name".

**Comment by solpeth** [comment]
Update: `parseFriendlyPackage` now seems to work. This PR should be ready for merging.

I've had to write a custom parser instead of using the `parsec` method from the `Parsec PackageName` instance, which fails an entire parse if it encounters version numbers. This custom parser _technically_ doesn't comply entirely with the Cabal spec, since it will not parse a package whose name includes leading digits after a hyphen. For example, let's apply parseCabalPackageName to the following inputs:

```haskell
"foo" -- pass
"foo1" -- pass
"foo-bar" -- pass
"foo-bar123" -- pass
"foo-bar-1" -- pass. Parser backtracks upon hitting the dash followed by the digit, returning only "foo-bar"
"foo-bar-1baz" -- FAIL. Parser will backtrack upon hitting the dash followed by the digit, assuming the rest of the string is a version.
```

To my knowledge there are no packages in `::haskell` which will fail this parser, but I will work on the parser over time to make it genuinely conform with the Cabal spec.

**To summarise the overall changes of this pull request:**

- Port pretty-printing away from `Distribution.Text` to `Distribution.Pretty`
- Port parsing away from `Distribution.Compat.ReadP` to `Distribution.Parsec.Class`[1] and `Distribution.Compat.CharParsing`
- Should the `parseFriendlyPackage` function fail, retain the underlying error message from Parsec and display it as a `HackPortError` to assist with diagnosing parsing faults.
- Use `simpleParsec` instead of the in-house `parseVersion` parser, which was an unnecessary duplication of the same functionality.

Requesting another review as I am satisfied that the correct functionality should now be present.

**Comment by solpeth** [comment]
@trofi thank you. I will make a couple of minor changes and push within a few hours.
